File: Segundo/Segundo_Cuatrimestre/Informatica/PDOO/SUBIR/Examen19-20/Hangar.rb
require_relative 'HangarToUI'
require_relative 'Weapon'
require_relative 'ShieldBooster'
require_relative 'WeaponType'

module Deepspace
	class Hangar
		def initialize(capacity)  # int 
			@maxElements = capacity
			@shieldBoosters = Array.new()
			@weapons = Array.new()
		end
		
		attr_reader :maxElements, :weapons, :shieldBoosters
		
		def self.newCopy(h)
			
			copy=new(h.maxElements)
			
			h.shieldBoosters.each do |shieldBooster|
				copy.addShieldBooster(shieldBooster)
			end
			
			h.weapons.each do |weapon|
				copy.addWeapon(weapon)
			end
			
			return copy
		end
		
		def spaceAvailable()
			retorno=true
						
			if (@maxElements <= (@weapons.length + @shieldBoosters.length))
				retorno=false
			end
			
			return (retorno)
		end
		
		def addWeapon(w)    # donde w es un Weapon
			retorno=false
			
			if spaceAvailable()
				retorno=true
				@weapons << w
			end
			
			return retorno       # devuelve un bool en función de si se puede añadir
		end
		
		def addShieldBooster(s)    # donde s es un ShieldBooster
			retorno=false
						
			if spaceAvailable()
				retorno=true
				@shieldBoosters << s 
			end
			  
			return retorno    # devuelve un bool en función de si se puede añadir
		end
		
		def removeWeapon(w)	   # donde w es un entero (indice de array)
					
			if (w < @weapons.length && w >= 0)
				retorno = @weapons.delete_at(w)
			else
				retorno = nil
			end
			
			return retorno   # devuelve el weapon eliminado
		end
		
		def  removeShieldBooster(s)    # donde s es un entero (indice de array)
			
			if (s < @shieldBoosters.length && s >= 0)
				retorno = @shieldBoosters.delete_at(s)
			else
				retorno = nil
			end
			
			return retorno     # devuelve el shield eliminado
		end
				
		def getUIversion()
			return HangarToUI.new(self)
		end
		
		def to_s
			out="Hangar - capacity: #{@maxElements}\n"
			out+="\tWeapons: [#{@weapons.join(' // ')}]\n"
			out+="\tShieldBoosters: [#{@shieldBoosters.join(' // ')}]\n"
			out+="------- end of Hangar -------"
			return out
		end

	end
end
		
		
			
			

File: Segundo/Segundo_Cuatrimestre/Informatica/PDOO/SUBIR/MARCIANITOS_P3/Marcianitos/src/deepspace/Hangar.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package deepspace;

import java.util.ArrayList;

/**
 *
 * @author laura
 */
public class Hangar {
    private int maxElements;
    private ArrayList<ShieldBooster> shieldBoosters; 
    private ArrayList<Weapon> weapons;
    
    Hangar(int capacity){
        this.maxElements = capacity;
        this.shieldBoosters = new ArrayList<ShieldBooster>();
        this.weapons = new ArrayList<Weapon>();
    }
    
    Hangar(Hangar h) {
        this.maxElements = h.maxElements;
        this.shieldBoosters = h.getShieldBoosters();
        this.weapons = h.getWeapons();
    }
    
    HangarToUI getUIversion() {
        return new HangarToUI(this);
    }
    
    private boolean spaceAvailable(){
        return (getMaxElements() > (getShieldBoosters().size() + getWeapons().size()));
    }
    
    public boolean addWeapon(Weapon w) {
        boolean added = false;
        if (spaceAvailable()) {
            getWeapons().add(w);
            added=true;
        }
        return added;
    }
    
    public boolean addShieldBooster(ShieldBooster s) {
        boolean added = false;
        if (spaceAvailable()) {
            getShieldBoosters().add(s);
            added=true;
        }
        return added;
    }
    
    public int getMaxElements() {
        return this.maxElements;
    }
    
    public ArrayList<ShieldBooster> getShieldBoosters(){
        return this.shieldBoosters;
    }
    
    public ArrayList<Weapon> getWeapons() {
        return this.weapons;
    }
    
    public ShieldBooster removeShieldBooster(int s){
        ShieldBooster shield = null;
        if (s < getShieldBoosters().size() && s >= 0) {
            shield = getShieldBoosters().get(s);
            getShieldBoosters().remove(s);
        }
        return shield;
    }
    
    public Weapon removeWeapon(int w){
        Weapon weapon = null;
        if (w < getWeapons().size() && w >= 0) {
            weapon = getWeapons().get(w);
            getWeapons().remove(w);
        }
        return weapon;    
    }
    
}
